ngraph/core/reference/src/runtime/reference/experimental_detectron_proposal_single_image.cpp

```diff
@@ -0,0 +1,381 @@
+                const float height = y1 - y0 + coordinates_offset;
+                const float area = std::max(0.0f, width) * std::max(0.0f, height);
+
+                const float b_width = x1j - x0j;
+                const float b_height = y1j - y0j;
+                const float b_area =
+                    (b_width + coordinates_offset) * (b_height + coordinates_offset);
+
+                const float intersection_area = area / (a_area + b_area - area);
+
+                is_dead[tail] = (nms_thresh < intersection_area) && (x0i <= x1j) && (y0i <= y1j) &&
+                                (x0j <= x1i) && (y0j <= y1i);
+            }
+        }
+        *num_out = count;
+    }
+
+    void fill_output_blobs(const float* proposals,
+                           const int64_t* roi_indices,
+                           float* rois,
+                           float* scores,
+                           const int64_t num_proposals,
+                           const int64_t num_rois,
+                           const int64_t post_nms_topn)
+    {
+        const float* src_x0 = proposals + 0 * num_proposals;
+        const float* src_y0 = proposals + 1 * num_proposals;
+        const float* src_x1 = proposals + 2 * num_proposals;
+        const float* src_y1 = proposals + 3 * num_proposals;
+        const float* src_score = proposals + 4 * num_proposals;
+
+        for (int64_t i = 0; i < num_rois; ++i)
+        {
+            int64_t index = roi_indices[i];
+            rois[i * 4 + 0] = src_x0[index];
+            rois[i * 4 + 1] = src_y0[index];
+            rois[i * 4 + 2] = src_x1[index];
+            rois[i * 4 + 3] = src_y1[index];
+            scores[i] = src_score[index];
+        }
+
+        if (num_rois < post_nms_topn)
+        {
+            for (int64_t i = 4 * num_rois; i < 4 * post_nms_topn; i++)
+            {
+                rois[i] = 0.0f;
+            }
+            for (int64_t i = num_rois; i < post_nms_topn; i++)
+            {
+                scores[i] = 0.0f;
+            }
+        }
+    }
+} // namespace
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            void experimental_detectron_proposals_single_image(
+                const float* im_info,
+                const float* anchors,
+                const float* deltas,
+                const float* scores,
+                const op::v6::ExperimentalDetectronGenerateProposalsSingleImage::Attributes& attrs,
+                const Shape& im_info_shape,
+                const Shape& anchors_shape,
+                const Shape& deltas_shape,
+                const Shape& scores_shape,
+                float* output_rois,
+                float* output_scores)
+            {
+                const int64_t anchors_num = static_cast<int64_t>(scores_shape[0]);
+
+                // bottom shape: (num_anchors) x H x W
+                const int64_t bottom_H = static_cast<int64_t>(deltas_shape[1]);
```

Comment:
Is it correct that we cast `float` to `int64_t`? Why do we use float for original values?

Reply:
`deltas_shape` has the type `Shape`. Hence, `deltas_shape[1]` has the type `size_t`, not `float`.